wrappers/services_test.go

```diff
@@ -223,7 +223,10 @@ func (s *servicesTestSuite) TestStartServices(c *C) {
 	err := wrappers.StartServices(info.Services(), nil)
 	c.Assert(err, IsNil)
 
-	c.Assert(s.sysdLog, DeepEquals, [][]string{{"start", filepath.Base(svcFile)}})
+	c.Assert(s.sysdLog, DeepEquals, [][]string{
+		{"--root", s.tempdir, "is-enabled", filepath.Base(svcFile)},
+		{"start", filepath.Base(svcFile)},
```

Comment:
It'd be nice to add a test for a case when the service has been disabled by the hook.

I've uploaded a small diff with all the changes here: https://paste.ubuntu.com/p/YzymnqMgMw/

Reply:
There is an unrelated problem with unit tests currently and service ordering, I'll fix that in master.

Reply:
Do you need me to add the `NewError` function you have in that diff?

Reply:
Got it, this is done.

Reply:
If you actually mock the systemctl command (there are examples of how to do this elsewhere), you can make it return the exact output necessary and the exact error code.  That's better than faking an internal error which might not be generated properly depending on bugs elsewhere.

Reply:
@niemeyer Actually, even if you use `MockSystemctl`, you still need to access the private structs of the `Error` struct when returning an error so you still don't avoid that problem. The only way way I see to avoid this is to change `systemctlCmd` not return an `*systemd.Error` and instead return something more generic that can be override outside the `systemd` package. 
The reason we have this problem is because we're testing things coming from the `systemd` package inside the `wrappers` package.

Reply:
@anonymouse64 this is the alternative suggested by @niemeyer:

```
diff --git a/systemd/systemd.go b/systemd/systemd.go
index 05e5f36d7..25f43748c 100644
--- a/systemd/systemd.go
+++ b/systemd/systemd.go
@@ -53,7 +53,7 @@ var systemctlCmd = func(args ...string) ([]byte, error) {
        bs, err := exec.Command("systemctl", args...).CombinedOutput()
        if err != nil {
                exitCode, _ := osutil.ExitCode(err)
-               return nil, NewError(exitCode, args, bs)
+               return nil, &Error{exitCode: exitCode, cmd: args, msg: bs}
        }
 
        return bs, nil
@@ -369,11 +369,6 @@ type Error struct {
        exitCode int
 }
 
-// NewError creates a systemd.Error, exported for mocks during testing
-func NewError(exitCode int, cmd []string, msg []byte) *Error {
-       return &Error{exitCode: exitCode, cmd: cmd, msg: msg}
-}
-
 func (e *Error) Error() string {
        return fmt.Sprintf("%v failed with exit status %d: %s", e.cmd, e.exitCode, e.msg)
 }
diff --git a/wrappers/services_test.go b/wrappers/services_test.go
index 5fdb1b8f3..88a0785ae 100644
--- a/wrappers/services_test.go
+++ b/wrappers/services_test.go
@@ -233,24 +233,34 @@ func (s *servicesTestSuite) TestNoStartDisabledServices(c *C) {
        info := snaptest.MockSnap(c, packageHello, &snap.SideInfo{Revision: snap.R(12)})
        svcFile := filepath.Join(s.tempdir, "/etc/systemd/system/snap.hello-snap.svc1.service")
 
-       r := systemd.MockSystemctl(func(cmd ...string) ([]byte, error) {
-               s.sysdLog = append(s.sysdLog, cmd)
-               sdcmd := cmd[0]
-               if len(cmd) >= 2 {
-                       sdcmd = cmd[len(cmd)-2]
-               }
-               if sdcmd == "is-enabled" {
-                       return nil, systemd.NewError(1, cmd, []byte("disabled\n"))
-               }
-               panic("unexpected systemctl command " + sdcmd)
-       })
-       defer r()
+       s.restorer()
+       r := testutil.MockCommand(c, "systemctl", `#!/bin/sh
+if [ "$1" = "--root" ]; then
+    shift 2
+fi
+
+case "$1" in
+    is-enabled)
+        if [ "$2" = "snap.hello-snap.svc1.service" ]; then
+            echo "disabled"
+            exit 1
+        else
+            echo "unexpected call $*"
+            exit 2
+        fi
+        ;;
+    *)
+        echo "unexpected call $*"
+        exit 2
+esac
+`)
+       defer r.Restore()
 
        err := wrappers.StartServices(info.Services(), nil)
        c.Assert(err, IsNil)
 
-       c.Assert(s.sysdLog, DeepEquals, [][]string{
-               {"--root", s.tempdir, "is-enabled", filepath.Base(svcFile)},
+       c.Assert(r.Calls(), DeepEquals, [][]string{
+               {"systemctl", "--root", s.tempdir, "is-enabled", filepath.Base(svcFile)},
        })
 }
 
```

Can you apply this and drop `NewError()` ?

Reply:
I see thanks for explaining. I'll add this shortly.